src/distribution/cauchy.rs

```diff
@@ -0,0 +1,232 @@
+}
+
+impl distribution::Inverse for Cauchy {
+    #[inline]
+    fn inverse(&self, p: f64) -> f64 {
+        use std::f64::{consts::PI, INFINITY, NEG_INFINITY};
+
+        should!((0.0..=1.0).contains(&p));
+
+        if p <= 0.0 {
+            NEG_INFINITY
+        } else if 1.0 <= p {
+            INFINITY
+        } else {
+            self.x_0 + self.gamma * (PI * (p - 0.5)).tan()
+        }
+    }
+}
+
+impl distribution::Median for Cauchy {
+    #[inline]
+    fn median(&self) -> f64 {
+        self.x_0
+    }
+}
+
+impl distribution::Modes for Cauchy {
+    #[inline]
+    fn modes(&self) -> Vec<f64> {
+        vec![self.x_0]
+    }
+}
+
+impl distribution::Sample for Cauchy {
+    #[inline]
+    fn sample<S>(&self, source: &mut S) -> f64
+    where
+        S: Source,
+    {
+        let gaussian = distribution::Gaussian::new(0.0, 1.0);
+        let a = gaussian.sample(source);
+        let b = gaussian.sample(source);
+        self.x_0() + self.gamma() * a / (b.abs() + f64::EPSILON)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use assert;
+    use prelude::*;
+
+    macro_rules! new(
+        ($x_0:expr, $gamma:expr) => (Cauchy::new($x_0, $gamma));
+    );
+
+    #[test]
+    fn density() {
+        let d = new!(2.0, 8.0);
+        let x = vec![-1.0, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 4.0, 6.0, 12.0];
+        let p = vec![
+            0.03488327519822364,
+            0.03744822190397538,
+            0.03843742021842001,
+            0.039176601376466544,
+            0.03963391578942141,
+            0.039788735772973836,
+            0.03963391578942141,
+            0.039176601376466544,
+            0.03744822190397538,
+            0.03183098861837907,
+            0.015527311521160521,
+        ];
+
+        assert::close(
+            &x.iter().map(|&x| d.density(x)).collect::<Vec<_>>(),
+            &p,
+            1e-15,
+        );
+    }
+
+    #[test]
+    fn distribution() {
+        let d = new!(2.0, 8.0);
+        let x = vec![
+            -1.0, 0.0, 0.01, 0.05, 0.1, 0.15, 0.25, 0.5, 1.0, 1.5, 2.0, 3.0, 4.0,
+        ];
+        let p = vec![
+            0.3857997487800918,
+            0.4220208696226307,
+            0.4223954618429798,
+            0.4238960166273086,
+            0.4257765641957529,
+            0.42766240385764065,
+            0.43144951512041,
+            0.44100191513247144,
+            0.46041657583943446,
+            0.48013147569445913,
+            0.5,
+            0.5395834241605656,
+            0.5779791303773694,
+        ];
+
+        assert::close(
+            &x.iter().map(|&x| d.distribution(x)).collect::<Vec<_>>(),
+            &p,
+            1e-15,
+        );
+    }
+
+    #[test]
+    fn entropy() {
+        use std::f64::consts::PI;
+        assert_eq!(new!(2.0, 1.0).entropy(), (PI * 4.0).ln());
+        assert::close(new!(3.0, 5.2).entropy(), 4.1796828725566719243, 1e-15);
+    }
+
+    #[test]
+    fn inverse() {
+        let d = new!(2.0, 3.0);
+        let x = vec![
```

Comment:
Could you please add both infinities to this test?

Reply:
Done.